### PR TITLE
8327365: Inline and remove GCStats

### PR DESCRIPTION
**Pull request by albertnetymk**
Simple refactoring to move a field from superclass to subclass.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8327365](https://bugs.openjdk.org/browse/JDK-8327365): Inline and remove GCStats (**Enhancement** - P4)

### Reviewers
 * [Guoxiong Li](https://openjdk.org/census#gli) (@lgxbslgx - Committer)
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18122/head:pull/18122` \
`$ git checkout pull/18122`

Update a local copy of the PR: \
`$ git checkout pull/18122` \
`$ git pull https://git.openjdk.org/jdk.git pull/18122/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18122`

View PR using the GUI difftool: \
`$ git pr show -t 18122`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18122.diff">https://git.openjdk.org/jdk/pull/18122.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18122#issuecomment-1978586726)

### CODE REVIEW

File: src/hotspot/share/gc/serial/generation.hpp [hunk 1]
```diff
@@ -168,13 +165,6 @@ class Generation: public CHeapObj<mtGC> {
...
   virtual HeapWord* expand_and_allocate(size_t word_size, bool is_tlab) = 0;
 
-  // Generations may keep statistics about collection. This method
-  // updates those statistics. current_generation is the generation
-  // that was most recently collected. This allows the generation to
-  // decide what statistics are valid to collect. For example, the
-  // generation can decide to gather the amount of promoted data if
-  // the collection of the young generation has completed.
-  GCStats* gc_stats() const { return _gc_stats; }
   virtual void update_gc_stats(Generation* current_generation, bool full) {}
```
Comment (lgxbslgx) [inline src/hotspot/share/gc/serial/generation.hpp:168]: The method `update_gc_stats` here can be removed and the method `update_gc_stats` in `tenuredGeneration.hpp` can be non-virtual.

**Review by lgxbslgx — CHANGES_REQUESTED** [review]
Looks good.

**Review by lgxbslgx — APPROVED** [review]
Looks good.

**Comment by tschatzl** [comment]
There is no more `GCStats` use in serial gc, all references to it in `generation.hpp` and `tenuredGeneration.hpp` can be removed (forward decl, include).
Also `GCStats` files can be moved to the `parallel` directory.

Tbh, similar to serial gc, `GCStats` could be replaced by the `AdaptivePaddedNoZeroDevAverage` class in Parallel GC too and completely removed. I.e. imo either both or none of these replacements should be done.

**Comment by albertnetymk** [comment]
I think `GCStats` can be inlined in Parallel as well. That is independent of this PR (Serial-only), so I left it out here.

**Review by lgxbslgx — APPROVED** [review]

**Comment by albertnetymk** [comment]
Included both Serial and Parallel in the latest revision.

**Review by tschatzl — APPROVED** [review]
Lgtm.

`_avg_promoted` does not need to be a reference I think, but that can be changed separately potentially together with the other statistics.

**Comment by albertnetymk** [comment]
Thanks for review.

/integrate

**Pull request closed without merging**